File: 18_10_2024_Count_Number_of_Maximum_Bitwise-OR_Subsets.java
import java.util.List;

class Solution {
    public int solve(int i, int ors, int maxi, List<Integer> nums) {
        if (i >= nums.size()) {
            return (ors == maxi) ? 1 : 0;
        }

        int cnt = 0;
        cnt += solve(i + 1, ors | nums.get(i), maxi, nums);
        cnt += solve(i + 1, ors, maxi, nums);

        return cnt;
    }

    public int countMaxOrSubsets(List<Integer> nums) {
        int maxi = 0;

        for (int it : nums) {
            maxi |= it;
        }

        return solve(0, 0, maxi, nums);
    }
}

File: 18_10_2024_Count_Number_of_Maximum_Bitwise-OR_Subsets.py
class Solution:
    def solve(self, i, ors, maxi, nums):
        if i >= len(nums):
            return 1 if ors == maxi else 0

        cnt = 0
        cnt += self.solve(i + 1, ors | nums[i], maxi, nums)
        cnt += self.solve(i + 1, ors, maxi, nums)

        return cnt

    def countMaxOrSubsets(self, nums):
        maxi = 0
        
        for it in nums:
            maxi |= it

        return self.solve(0, 0, maxi, nums)

